services/QuillLMS/app/models/feedback_history.rb

```diff
@@ -149,7 +157,52 @@ def self.batch_create(param_array)
     self.feedback_session_uid = FeedbackSession.get_uid_for_activity_session(feedback_session_uid)
   end
 
-  def self.list_by_activity_session(activity_id: nil, page: 1, start_date: nil, end_date: nil, page_size: DEFAULT_PAGE_SIZE, turk_session_id: nil)
+  def self.apply_activity_session_filter(query, filter_type)
+    case filter_type
+    when FILTER_ALL
+      query
+    when FILTER_SCORED
+      query = query.where("feedback_history_ratings.rating IS NOT NULL")
+    when FILTER_UNSCORED
+      query = query.where("feedback_history_ratings.rating IS NULL")
+    when FILTER_WEAK
+      query = query.where("feedback_history_ratings.rating IS FALSE")
+    when FILTER_COMPLETE
+      query = query.having(
+        <<-SQL
+        (
+          CASE WHEN
+            ((COUNT(CASE WHEN comprehension_prompts.conjunction = 'because' AND feedback_histories.optimal THEN 1 END) = 1) OR
+              (COUNT(CASE WHEN comprehension_prompts.conjunction = 'because' THEN 1 END) = MAX(CASE WHEN comprehension_prompts.conjunction = 'because' THEN comprehension_prompts.max_attempts END))) AND
+            ((COUNT(CASE WHEN comprehension_prompts.conjunction = 'but' AND feedback_histories.optimal THEN 1 END) = 1) OR
+              (COUNT(CASE WHEN comprehension_prompts.conjunction = 'but' THEN 1 END) = MAX(CASE WHEN comprehension_prompts.conjunction = 'but' THEN comprehension_prompts.max_attempts END))) AND
+            ((COUNT(CASE WHEN comprehension_prompts.conjunction = 'so' AND feedback_histories.optimal THEN 1 END) = 1) OR
+              (COUNT(CASE WHEN comprehension_prompts.conjunction = 'so' THEN 1 END) = MAX(CASE WHEN comprehension_prompts.conjunction = 'so' THEN comprehension_prompts.max_attempts END)))
+          THEN true ELSE false END
+        ) = true
+        SQL
+      )
+    when FILTER_INCOMPLETE
+      query = query.having(
+        <<-SQL
+        (
+          CASE WHEN
+            ((COUNT(CASE WHEN comprehension_prompts.conjunction = 'because' AND feedback_histories.optimal THEN 1 END) = 1) OR
+              (COUNT(CASE WHEN comprehension_prompts.conjunction = 'because' THEN 1 END) = MAX(CASE WHEN comprehension_prompts.conjunction = 'because' THEN comprehension_prompts.max_attempts END))) AND
+            ((COUNT(CASE WHEN comprehension_prompts.conjunction = 'but' AND feedback_histories.optimal THEN 1 END) = 1) OR
+              (COUNT(CASE WHEN comprehension_prompts.conjunction = 'but' THEN 1 END) = MAX(CASE WHEN comprehension_prompts.conjunction = 'but' THEN comprehension_prompts.max_attempts END))) AND
+            ((COUNT(CASE WHEN comprehension_prompts.conjunction = 'so' AND feedback_histories.optimal THEN 1 END) = 1) OR
+              (COUNT(CASE WHEN comprehension_prompts.conjunction = 'so' THEN 1 END) = MAX(CASE WHEN comprehension_prompts.conjunction = 'so' THEN comprehension_prompts.max_attempts END)))
+          THEN true ELSE false END
+        ) = false
+        SQL
+      )
```

Comment:
I think these two are only differing in the boolean at the very end, right? Maybe we could just pass that boolean to a function which will return the whole SQL string instead of repeating the rest of it? I think it also probably makes sense to constantize the various conjunctions in here.

Reply:
good suggestions! I just updated